engine/verification/fetcher/engine.go

```diff
@@ -106,352 +102,209 @@ func (e *Engine) Done() <-chan struct{} {
-// its chunk data pack.
-// it also retries the chunk data request if the data hasn't been received
-// for a while.
-func (e *Engine) onTimer() {
-	allChunks := e.pendingChunks.All()
 
-	e.log.Debug().Int("total", len(allChunks)).Msg("start processing all pending pendingChunks")
+	lg = lg.With().
+		Uint64("chunk_index", status.Chunk.Index).
+		Hex("block_id", logging.ID(status.Chunk.BlockID)).
+		Hex("result_id", logging.ID(status.ExecutionResultID)).Logger()
 
-	sealed, err := e.state.Sealed().Head()
+	// make sure the chunk data pack is valid
+	err := e.validateChunkDataPack(status.Chunk, originID, chunkDataPack, collection)
 	if err != nil {
-		e.log.Error().Err(err).Msg("could not get last sealed block")
+		// TODO: this can be due to a byzantine behavior
+		lg.Error().Err(err).Msg("could not validate chunk data pack")
 		return
 	}
 
-	allExecutors, err := e.state.Final().Identities(filter.HasRole(flow.RoleExecution))
+	result, err := e.results.ByID(status.ExecutionResultID)
 	if err != nil {
-		e.log.Error().Err(err).Msg("could not get executors")
+		// potential database leakage or corruption.
+		lg.Fatal().Err(err).Msg("could not retrieve execution result of chunk status, possibly a bug")
 		return
 	}
 
-	for _, chunk := range allChunks {
-		chunkID := chunk.ID()
+	lg = lg.With().
+		Hex("result_id", logging.ID(result.ID())).
+		Hex("block_id", logging.ID(status.Chunk.BlockID)).Logger()
 
-		lg := e.log.With().
-			Hex("chunk_id", logging.ID(chunkID)).
-			Hex("result_id", logging.ID(chunk.ExecutionResultID)).
-			Hex("block_id", logging.ID(chunk.Chunk.ChunkBody.BlockID)).
-			Uint64("height", chunk.Height).
-			Logger()
-
-		// if block has been sealed, then we can finish
-		isSealed := chunk.Height <= sealed.Height
-
-		if isSealed {
-			removed := e.pendingChunks.Rem(chunkID)
-			lg.Info().Bool("removed", removed).Msg("chunk has been sealed, no longer needed")
-			continue
-		}
-
-		// check if has reached max try
-		if !CanTry(e.maxAttempt, chunk) {
-			lg.Debug().
-				Int("max_attempt", e.maxAttempt).
-				Int("actual_attempts", chunk.Attempt).
-				Msg("max attempts reached, no longer fetch data pack for chunk")
-			continue
-		}
-
-		err := e.requestChunkDataPack(chunk, allExecutors)
-		if err != nil {
-			lg.Warn().Err(err).Msg("could not request chunk data pack")
-			continue
-		}
+	removed := e.pendingChunks.Rem(chunkDataPack.ChunkID)
+	lg.Debug().Bool("removed", removed).Msg("removed chunk status")
+	if !removed {
+		// we deduplicate the chunk data responses at this point, reaching here means a
+		// duplicate chunk data response is under process concurrently, so we give up
+		// on processing current one.
+		return
+	}
 
-		log.Info().Msg("chunk data requested")
+	// pushes chunk data pack to verifier, and waits for it to be verified.
+	err = e.pushToVerifier(status.Chunk, result, chunkDataPack, collection)
+	if err != nil {
+		lg.Fatal().Err(err).Msg("could not push the chunk to verifier engine")
+		return
```

Comment:
Despite `fatal` log level, we `return` to make sure that changing the log level does not change the logic.